ENH: add hermitian=False kwarg to np.linalg.matrix_rank

perimosocordiae opened the pull request:
With a symmetric matrix, the more efficient `eigvalsh` method can be used to find singular values.

This PR also make some minor docstring fixes, and uses the more efficient `count_nonzero` method.

perimosocordiae committed:
ENH: add hermitian=False kwarg to matrix_power

With a symmetric matrix, the more efficient `eigvalsh` method
can be used to find singular values.

seberg commented on numpy/linalg/linalg.py:
```diff
@@ -1504,22 +1504,25 @@ def cond(x, p=None):
        ``eps`` is the epsilon value for datatype of ``S``, then `tol` is
        set to ``S.max() * max(M.shape) * eps``.
+    symmetric : bool, optional
+       If True (default), `M` is assumed to be square and symmetric, enabling
```
its not the default.

seberg commented:
More tests could be nice, such as also for the tolerance. Which also makes me wonder, the singular values are not quite the eigenvalues, no? So if you would change to symmetric, the difference, with respect to the tolerance setting, should be more then just numerical accuracy?

perimosocordiae commented:
Should I make a new method for the symmetric option tests or keep them where they are now?

My last commit fixes the discrepancy by using the absolute value of the eigenvalues, which (as I understand the math) exactly matches the singular values for a symmetric matrix:

 * `svd(X)` = `sqrt(eigvals(X.T.dot(X)))` in the general case
 * `X.T.dot(X)` = `X.dot(X)` = `matrix_power(X, 2)` in the symmetric case
 * `eigvals(matrix_power(X, k))` = `eigvals(X)**k` in general

Combining these equalities gives us `svd(X)` = `sqrt(eigvals(X)**2)` = `abs(eigvals(X))`.

seberg commented:
Yeah, probably right for symmetry. Do whatever you feel looks nicer, my guess is, it would look nicer in different functions ;) and making tests shorter makes it easier to find the code that broke if it gets triggered.

charris commented:
What is the speed/accuracy advantage?

charris commented:
Isn't the title of this PR incorrect? Looks like matrix_rank is the relevant function. Same with the summary line of the first commit.

perimosocordiae commented:
@charris: Oops, you're correct on the title mixup. My bad.

Here's some unscientific testing on my Ubuntu machine with master-branch numpy:
```
In [1]: a = np.random.random((1000, 1000))

In [2]: a = (a + a.T)/2

In [3]: %timeit np.linalg.svd(a, compute_uv=False)
1 loop, best of 3: 218 ms per loop

In [4]: %timeit np.abs(np.linalg.eigvalsh(a))
10 loops, best of 3: 128 ms per loop
```

perimosocordiae commented:
I'm not sure how to tell which method is more numerically accurate. They do produce very slightly different results on large badly-scaled matrices, but it's unclear which introduces more floating-point error.

charris commented:
How about the name `ishermitean` instead?

perimosocordiae commented:
@charris I changed it to `is_hermitian`.

homu commented:
:umbrella: The latest upstream changes (presumably #8368) made this pull request unmergeable. Please resolve the merge conflicts.

eric-wieser commented:
Would be better as a rebase rather than a merge. Also, this will conflict with #8682 anyway

eric-wieser commented:
Given we already have `eig` + `eigh`, `eigvals` + `eigvalsh`, I think we should either:

 * Call the hermitian version `matrix_rankh`
 * Deprecate `eigh` and `eigvalsh` and add `eig(..., hermitian=True)`, `eigvals(..., hermitian=True)`

perimosocordiae commented:
Yeah, I was trying out Github's web-based conflict resolution tool. It's nice, but a rebase would indeed be cleaner. If the Github folks are watching, maybe that would be a nice feature to add? ;-) I'll rebase and squash when/if this PR is accepted for merging, especially because I need to fix my commit message.

I'm -1 for `matrix_rankh`, and in general I found the `-h` suffix versions pretty inscrutable back when I was first learning the API. Actually deprecating `eigh` and friends seems like it would affect a lot of code in the wild, though...

perimosocordiae commented:
One benefit of adding a `hermitian` kwarg would be to eventually enable `eig(..., hermitian='autodetect')` or similar, which could be convenient for some users and also matches the Matlab behavior.

homu commented:
:umbrella: The latest upstream changes (presumably #8682) made this pull request unmergeable. Please resolve the merge conflicts.

eric-wieser commented:
Needs rebase

perimosocordiae commented:
Rebased and squashed, and updated the title.

eric-wieser commented on numpy/linalg/linalg.py:
```diff
@@ -1568,10 +1572,13 @@ def matrix_rank(M, tol=None):
+    if hermitian:
+        S = eigvalsh(M)
+    else:
+        S = abs(svd(M, compute_uv=False))
```
Why did this `abs` appear? Should it be in the other branch?

perimosocordiae replied:
See [this comment](https://github.com/numpy/numpy/pull/8557#issuecomment-277075209) for the derivation.

eric-wieser replied:
Yes, I think you should have `abs(eigvalsh(X))` on the line above. Why are you changing what happens for `hermitian = False`? Was there a bug?

perimosocordiae replied:
Oof, my bad! Fixed now.

eric-wieser commented on numpy/linalg/linalg.py:
```diff
@@ -1489,22 +1489,26 @@ def cond(x, p=None):
+    hermitian : bool, optional
+       If True, `M` is assumed to be Hermitian (symmetric if real-valued),
+       enabling the use of a more efficient method for finding singular values.
+       Defaults to False.
```
Needs `.. versionadded:: 1.14`

perimosocordiae replied:
Done.

eric-wieser commented:
An argument for `matrix_rankh`: we could then add the `UPLO` argument, and forward it to `eigvalsh`

perimosocordiae commented:
I'm not sure how useful allowing `'U'` and `'L'` would be, so I'll defer to the consensus.

eric-wieser commented:
> I'm not sure how useful allowing 'U' and 'L' would be,

I agree, but we already expose it on `eigvals`, so there's an argument to keep it. I only mention it as this is probably going to trigger a PR to move all the functions to `hermitian=...` at some point, and it would be nice to have a roadmap now.

perimosocordiae commented:
I've opened gh-9436 to get that roadmap out explicitly.

eric-wieser commented:
Needs a release note.

Might be nice to add a benchmark comparing hermitian with not, but not necessary.

perimosocordiae commented:
Added a release note and rebased. I didn't add an asv benchmark because those are more useful for tracking efficiency improvements for the same codepaths. I did run a final sanity check using `runtests.py -i`, though:
```
Python 3.5.3 (default, Jan 19 2017, 14:11:04) 
Type 'copyright', 'credits' or 'license' for more information
IPython 6.1.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: x = np.random.random((500, 500))

In [2]: x = (x + x.T) / 2

In [3]: %timeit np.linalg.matrix_rank(x)
157 ms ± 194 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [4]: %timeit np.linalg.matrix_rank(x, hermitian=True)
46.7 ms ± 264 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

perimosocordiae commented:
Rebased to fix merge conflicts.

eric-wieser commented:
More conflicts, I'm afraid. I'll try and merge this before it conflicts again

perimosocordiae commented:
Conflicts resolved.

eric-wieser commented:
I don't agree with that last commit - that doesn't seem to match the style of other files.

Can you push again without that commit? Also, I don't think I can restart appveyor with a `ci-skip` commit last.

perimosocordiae commented:
Ok, I reverted that change and squashed the commits. Hopefully appveyor will work this time!

eric-wieser commented:
Oh I see - the last commit brought it in line with what it was before. Whatever, let's just put this in, since the tests now run. Thanks @perimosocordiae!

perimosocordiae commented:
Thanks for sticking with this, @eric-wieser !
